modulus/utils/graphcast/loss.py

```diff
@@ -110,3 +113,116 @@ def forward(self, invar: torch.Tensor, outvar: torch.Tensor) -> torch.Tensor:
         return CustomCellAreaWeightedLossAutogradFunction.apply(
             invar, outvar, self.area
         )
+
+
+class GraphCastLossFunction(nn.Module):
+    """Loss function as specified in GraphCast.
+    Parameters
+    ----------
+    area : torch.Tensor
+        Cell area with shape [H, W].
+    """
+
+    def __init__(self, area, channels_list, dataset_metadata_path, time_diff_std_path):
+        super().__init__()
+        self.area = area
+        self.channel_dict = self.get_channel_dict(dataset_metadata_path, channels_list)
+        self.variable_weights = self.assign_variable_weights()
+        self.time_diff_str = self.get_time_diff_str(time_diff_std_path, channels_list)
+
+    def forward(self, invar, outvar):
+        """
+        Implicit forward function which computes the loss given
+        a prediction and the corresponding targets.
+        Parameters
+        ----------
+        invar : torch.Tensor
+            prediction of shape [T, C, H, W].
+        outvar : torch.Tensor
+            target values of shape [T, C, H, W].
+        """
+
+        # outvar normalization
+        loss = (invar - outvar) ** 2  # [T,C,H,W]
+        # weighted by inverse variance
+        loss = (
+            loss
+            * 1.0
+            / torch.square(self.time_diff_str.view(1, -1, 1, 1).to(loss.device))
+        )
```

Comment:
Going back to the `time_diff` script, I strongly believe that the diffs should be weighted by area if that is also done here.

Reply:
Added area weighting